lyricsheets/models/karaoke.py

```diff
@@ -159,248 +166,217 @@ def middle(self) -> float:
-
-    @property
-    def right(self) -> float:
-        return self.left + self.width
+        for char in self.chars:
+            char.__dict__.pop("width", None)
+            char.__dict__.pop("left", None)
 
-    @property
-    def middle(self) -> float:
-        return self.top + self.height / 2
+        self._style = style
 
     @property
-    def bottom(self) -> float:
-        return self.top + self.height
+    def resX(self) -> int:
+        return self._resX
 
-    def calculate_char_offsets(self, style: Style, transitionDuration: timedelta):
-        self.calculate_char_fade_offsets(style, transitionDuration)
-        for k in self.kara:
-            k.calculate_char_kara_times(style)
+    @resX.setter
+    def resX(self, resX: int):
+        # Invalidate cached properties
+        for syl in self.syls:
+            syl.__dict__.pop("left", None)
 
-    def calculate_char_fade_offsets(self, style: Style, transitionDuration: timedelta):
-        fontScaler = FontScaler(style)
-
-        lineCharTimes = [
-            timedelta(milliseconds=m)
-            for m in fontScaler.split_by_rendered_width(
-                pyass.timedelta(transitionDuration).total_milliseconds(), self.text
-            )
-        ]
-
-        charAccTime = timedelta()
-        for char, lineCharTime in zip(self.chars, lineCharTimes):
-            char.fadeOffset = charAccTime
-            charAccTime += lineCharTime
+        for char in self.chars:
+            char.__dict__.pop("left", None)
 
-    def bind_style(self, style: Style):
-        self.style = style
-        for kara in self.kara:
-            kara.style = style
-            for char in kara.chars:
-                char.style = self.style
+        self._resX = resX
 
-        self._calculate_sizing()
-        self._calculate_positions()
+    @property
+    def transitionDuration(self) -> timedelta:
+        return self._transitionDuration
 
-    def _calculate_sizing(self):
-        if not self.style:
-            raise StyleNotBoundException()
+    @transitionDuration.setter
+    def transitionDuration(self, transitionDuration: timedelta):
+        # Invalidate cached properties
+        self.__dict__.pop("_charFadeOffsets", None)
 
-        fontScaler = FontScaler(self.style)
+        for syl in self.syls:
+            syl.__dict__.pop("_charKaraTimes", None)
 
-        self._width = fontScaler.get_length(self.text)
-        self._height = self.style.fontSize
+        self._transitionDuration = transitionDuration
 
-        for syl in self.kara:
-            syl._calculate_sizing()
+    @cached_property
+    def width(self) -> float:
+        return FontScaler(self.style).get_length(self.text)
 
-    def _calculate_positions(self, resX: int = 1920, resY: int = 1080):
-        if not self.style:
-            raise StyleNotBoundException()
+    @property
+    def height(self) -> float:
+        return self.style.fontSize
 
-        # Horizontal positioning
+    @property
```

Comment:
Any reason why this isn't cached?

Reply:
Mainly the reasoning of what is cached and what isn't is just how much work it takes to calculate something, so things that are cached are:
- Things that require FontScaler
- Things that require iterating through a list of things (e.g. the char offsets for one char is depending on all the preceding chars)